third_party/solara/README.md

```diff
@@ -0,0 +1,399 @@
+    text_retrieved = "\n\n".join(retrieved_chunk)
+    return text_retrieved
+```
+
+Finally, we edit `response_generator` to implement our new RAG with the files! This function, when there are PDFs, will extract the text with PyPDF2 and make use of `rag_pdf` to retrieve the relevant data. It will only then send the request to the model:
+
+```py
+def response_generator(messages: list, txt: List[str]):
+    response = client.chat_stream(
+        model = "open-mistral-7b",
+        messages = messages[:-1] + [{"role":"user","content": rag_pdf(txt, messages[-1]["content"]) + "\n\n" + messages[-1]["content"]}],
+        max_tokens = 1024
+    )
+    for chunk in response:
+        yield chunk.choices[0].delta.content
+```
+
+And everything is done! Now we can run our new interface with `solara run chat_with_pdfs.py`
+
+<details>
+<summary><b>chat_with_pdfs.py</b></summary>
+
+```py
+import io
+import solara as sl
+from mistralai.client import MistralClient
+import numpy as np
+import PyPDF2
+import faiss
+from solara.components.file_drop import FileInfo
+from solara.lab import use_task, Task
+from typing import List, cast
+from typing_extensions import TypedDict
+
+mistral_api_key = "your_api_key"
+client = MistralClient(api_key = mistral_api_key)
+
+def get_text_embedding(input_text: str):
+    embeddings_batch_response = client.embeddings(
+          model = "mistral-embed",
+          input = input_text
+      )
+    return embeddings_batch_response.data[0].embedding
+
+def rag_pdf(txt: List[str], question: str) -> str:
+    chunk_size = 1024
+    chunks = []
+    for _txt in txt:
+        chunks += [_txt[i:i + chunk_size] for i in range(0, len(_txt), chunk_size)]
+
+    text_embeddings = np.array([get_text_embedding(chunk) for chunk in chunks])
+    d = text_embeddings.shape[1]
+    index = faiss.IndexFlatL2(d)
+    index.add(text_embeddings)
+
+    question_embeddings = np.array([get_text_embedding(question)])
+    D, I = index.search(question_embeddings, k = 3)
+    retrieved_chunk = [chunks[i] for i in I.tolist()[0]]
+    text_retrieved = "\n\n".join(retrieved_chunk)
+    return text_retrieved
+
+class MessageDict(TypedDict):
+    role: str
+    content: str
+
+messages: sl.Reactive[List[MessageDict]] = sl.reactive([])
+
+def response_generator(messages: list, txt: List[str]):
+    response = client.chat_stream(
+        model = "open-mistral-7b", 
+        messages = messages[:-1] + [{"role":"user","content": rag_pdf(txt, messages[-1]["content"]) + "\n\n" + messages[-1]["content"]}],
+        max_tokens = 1024
+    )
+    for chunk in response:
+        yield chunk.choices[0].delta.content
+
+def add_chunk_to_ai_message(chunk: str):
+    messages.value = [
+        *messages.value[:-1],
+        {
+            "role": "assistant",
+            "content": messages.value[-1]["content"] + chunk,
+        },
+    ]
+
+@sl.component
+def Page():
+    with sl.Sidebar():
+        content, set_content = sl.use_state(cast(List[bytes], []))
```

Comment:
```suggestion
```
Not needed anymore, with the get_text task.
Also, I would not mix use_state and use_reactive, I would simply use one (I prefer use_reactive), it will only confuse new users I think.